src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -746,6 +751,37 @@ protected ComputedInstanceFieldLayout ComputeAutoFieldLayout(MetadataType type,
                 classLayoutSize: 0,
                 byteCount: out instanceByteSizeAndAlignment);
 
+            if (type.IsValueArray)
+            {
+                int repeat = type.GetValueArrayLength();
+
+                if (repeat <= 0)
+                {
+                    ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, type);
+                }
+
+                if (!instanceByteSizeAndAlignment.Size.IsIndeterminate)
+                {
+                    long size = instanceByteSizeAndAlignment.Size.AsInt;
+                    size *= repeat;
+
+                    // matching the max size validation in MethodTableBuilder
```

Comment:
We should limit this by intentionally chosen limit (e.g. 64 kB or 1MB) and not by this internal CoreCLR implementation detail.

Reply:
One possible use case is embedding in heap objects as inline storage - to avoid unnecessary indirections, so these will not necessarily be on stack and large instances could be useful. 
As an arbitrarily chosen limit, 64 kB would seem a bit small. We can raise the limit later, but I think we should start with something like 1MB.

Reply:
Also it may make sense to have the same limit in CoreCLR as well - to not cause surprises when an app switches to AOT.

Reply:
One of the larger "real world" fixed size buffers I'm aware of from the WinSDK is `1025 * 12` (just above 12kb) in size (`DXGI_GAMMA_CONTROL`).

The default windows stack size is itself 1MB. Unix defaults to 8MB. MacOS defaults to 8MB for the main thread and I believe 512kb for secondary threads.

For C, the minimum required "bytes in object" size is 65535, but MSVC/Clang/GCC all support larger.

Reply:
Could/should we just make this match `Array.MaxLength`. In practice most implementations may fail sooner due to memory constraints, but it at least makes it "consistent" and easy for users to determine.

Reply:
> Also we should have the same limit in CoreCLR as well - to not cause surprises when an app switches to AOT.

I would not worry about surprises like this. When people push against the system limits (e.g. stack limits), it is generally bad experience. We do not guarantee that these implementation limits are same between runtime versions, runtime modes, platforms or architectures.

Reply:
Since there are examples of existing 12Kb fixed buffers, I think 65K is within reach of some scenarios and thus could be constraining. 
I've picked 1MiB (1024 * 1024) as the limit. I think it should be beyond most plausible uses.